### ISSUE TITLE
Empty string with quotes.

### ISSUE THREAD

**Issue by edgbla** [ndevilla/iniparser#70]
for example:
filter = ""

this code can't handle it

```
} else if (sscanf (line, "%[^=] = \"%[^\"]\"", key, value) == 2
       ||  sscanf (line, "%[^=] = '%[^\']'",   key, value) == 2) {
```

this string will be processed in

```
} else if (sscanf (line, "%[^=] = %[^;#]", key, value) == 2) {
```

the condition

```
if (!strcmp(value, "\"\"") || (!strcmp(value, "''"))) {
        value[0]=0 ;
}
```

will not be processed.

result:
load = ""
save = """"
load = """"
save = """"""
etc.

**Comment by touilleMan** [comment]
Hi,

You're right, quotes are not well handled when a dictionary is resaved.

I've work a bit on this trouble and end up with a rewrite save/load mechanism (see https://github.com/ndevilla/iniparser/tree/quoting and 9daa5ac)
 - quotes (`"` and `'`) can be used to protect special values (`#`, `;`, `"`, `'`)
 - escaping `\` char can be used to protect special values outside quotes
 - escaping `\` char can also be used to represent `\n`, `\r`, `\t`
 - when saving a value, if it contains a special value it will be surrounded by `"`

According to wikipedia those behavior are part of some ini implementations (see https://en.wikipedia.org/wiki/INI_file#Escape_characters)

@ndevilla Given this change is not trivial I'd like to have your opinion on this before going further

**Comment by ndevilla** [comment]
Good catch @edgbla !
Seems the empty-quoted value is just not processed in the correct place. I would just move the test one case down, so that filter="" is understood like filter=
You would lose the empty quotes when saving back, but they were unneeded anyway, no?

@touilleMan I would rather avoid getting into character escape handling. Not so hard to write I believe, but it would break existing implementations that expect to see \n come in and out as \n (character backslash + character n) rather than a literal CR.
BTW: this is probably worth including in the documentation to avoid surprising users.

What do you think?

**Comment by touilleMan** [comment]
@ndevilla you're right about this : my fix break the api which is not really nice

I guess we can explicit the "no escaping" policy in the documentation and explain the user is expected to run `sprintf` on the returned value if he wants to cook the escaping (i.g. order to turn 2-chars literal `\n` into a real `\n`)

But I'm still not sure about the saving of a quoted value:

Considering an ini file
```
[main]
key = "   my    value   " # surounding witespaces
```
We load it into iniparser, then save the resulting dictionary:
```
[main]
key =    my   value   
```
This time the string `"   my    value   "` is become `   my value   ` (no quotes), which means if we reload it into iniparser the starting and ending witespaces will be stripped.

Solution could be to check on save if the value starts/ends with a space and then automatically surround it by quotes...

**Comment by edgbla** [comment]
I always use quotes for strings, they make the text more readable.
I can't use filter without quotes, spaces will be removed.

**Comment by touilleMan** [comment]
54716a77d621373ad3428170af6dacf9f106c264 solve the `key=""` bug, Better late then never ! :smile: 

@ndevilla What do you think about the point I exposed in my previous post about the way a quoted-string get it quotes removed when saved ? Do you see something to do about it ?

**Comment by ndevilla** [comment]
I would keep things the way they are, and document the fact that if you are using iniparser to read/write/read pathologic cases like `key="    value    "` you will lose the surrounding white spaces on the second read.

What makes ini a great config file format is the fact that it is so simple it can be understood without need to read some documentation.  Users can automatically guess how it works: a key is set to a value. The value is anything between the first and last non-blank character on the line, and comments are preceded by a hash.  Getting into whitespace handling when surrounded by quotes seems like opening a can of worms. You will have to tag such values as special cases in the parser to be able to print them back exactly as they were read. That can be done, but is it really worth it?
The initial goal of this library was to easily access values parsed from a config file, I never meant it to be a full-fledged round-trip editor. That said, I have nothing against extending the library to do such things. It would probably best be renamed as iniedit or similar to indicate its intentions.

**Comment by lmoellendorf** [comment]
> Good catch [@edgbla](https://github.com/edgbla) ! Seems the empty-quoted value is just not processed in the correct place. I would just move the test one case down, so that filter="" is understood like filter= You would lose the empty quotes when saving back, but they were unneeded anyway, no?

This had been fixed with commit 54716a77d621373ad3428170af6dacf9f106c264
